### PR TITLE
Ensure GitHub.TeamFoundation.15 MEF assembly is always installed on Visual Studio 2017

### LINKED ISSUE #1810: SetSite failed for the package [GitHubPackage], VS 15.8 Preview 4/5
This is a dupe of #1625 but the problem just recently started happening for me. GitHub extension is 2.5.4.3349. I'm on VS 2017 Professional, currently on 15.8 Preview 5. I believe the problem also happened with Preview 4. I get a popup error message every time I open a solution that uses git, whether it uses GitHub or not. The log is below.

```xml
<entry>
  <record>915</record>
  <time>2018/07/27 19:48:11.911</time>
  <type>Warning</type>
  <source>VisualStudio</source>
  <description>Unexpected system error mode before loading package [GitHubPackage]</description>
  <guid>{C3D3DC68-C977-411F-B3E8-03B0DCCF7DFC}</guid>
</entry>
<entry>
  <record>916</record>
  <time>2018/07/27 19:48:11.911</time>
  <type>Information</type>
  <source>VisualStudio</source>
  <description>Begin package load [GitHubPackage]</description>
  <guid>{C3D3DC68-C977-411F-B3E8-03B0DCCF7DFC}</guid>
</entry>
<entry>
  <record>917</record>
  <time>2018/07/27 19:48:11.953</time>
  <type>Error</type>
  <source>VisualStudio</source>
  <description>SetSite failed for package [GitHubPackage]Source: &apos;System.ComponentModel.Composition&apos; Description: No exports were found that match the constraint: &#x000A;&#x0009;ContractName&#x0009;GitHub.Commands.IOpenFromUrlCommand&#x000A;&#x0009;RequiredTypeIdentity&#x0009;GitHub.Commands.IOpenFromUrlCommand&#x000D;&#x000A;</description>
  <guid>{C3D3DC68-C977-411F-B3E8-03B0DCCF7DFC}</guid>
  <hr>80131500</hr>
  <errorinfo></errorinfo>
</entry>
<entry>
  <record>918</record>
  <time>2018/07/27 19:48:11.959</time>
  <type>Error</type>
  <source>VisualStudio</source>
  <description>End package load [GitHubPackage]</description>
  <guid>{C3D3DC68-C977-411F-B3E8-03B0DCCF7DFC}</guid>
  <hr>80131500</hr>
  <errorinfo></errorinfo>
</entry>
```

### LINKED ISSUE #1812: Error On Project Load
- GitHub Extension for Visual Studio version: 2.5.4.3349
- Visual Studio version: 15.0.26430.16
*NOTE: this is 15.2.6* https://docs.microsoft.com/en-us/visualstudio/install/visual-studio-build-numbers-and-release-dates

### when i load Project or solution vs shows alert with this message:

> An exception has been encountered. This may be caused by an extension.
> You can get more information by examining the file 'C:\Users\{user}\AppData\Roaming\Microsoft\VisualStudio\15.0_5de280e5\ActivityLog.xml'.

### and in gitub window shows:

> System.InvalidOperationException: Could not locate any instances of contract GitHub.Services.ITeamExplorerServiceHolder.
>    at GitHub.VisualStudio.GitHubServiceProvider.GetService(Type serviceType) in C:\projects\visualstudio\src\GitHub.VisualStudio\Services\GitHubServiceProvider.cs:line 147
>    at GitHub.VisualStudio.GitHubServiceProvider.GetService[T]() in C:\projects\visualstudio\src\GitHub.VisualStudio\Services\GitHubServiceProvider.cs:line 153
>    at GitHub.VisualStudio.UI.GitHubPane.<InitializeAsync>d__12.MoveNext() in C:\projects\visualstudio\src\GitHub.VisualStudio\UI\GitHubPane.cs:line 98

### LINKED ISSUE #1826: GitHub Extension broken with VS 15.7.6
- GitHub Extension for Visual Studio version: 2.5.4.3349
- Visual Studio version: 15.7.6

__What happened__ (with steps, logs and screenshots, if possible)

After updating to VS 15.7.6 I started getting an error every time I open VS

image

In the GitHub pane I see this error:

```
System.InvalidOperationException: Could not locate any instances of contract GitHub.Services.ITeamExplorerServiceHolder.
   at GitHub.VisualStudio.GitHubServiceProvider.GetService(Type serviceType) in C:\projects\visualstudio\src\GitHub.VisualStudio\Services\GitHubServiceProvider.cs:line 147
   at GitHub.VisualStudio.GitHubServiceProvider.GetService[T]() in C:\projects\visualstudio\src\GitHub.VisualStudio\Services\GitHubServiceProvider.cs:line 153
   at GitHub.VisualStudio.UI.GitHubPane.<InitializeAsync>d__12.MoveNext() in C:\projects\visualstudio\src\GitHub.VisualStudio\UI\GitHubPane.cs:line 98
```

### LINKED ISSUE #1859: Corrupt extension cache and missing MEF exports
- GitHub Extension for Visual Studio version: 2.5.4.3349 (probably any)
- Visual Studio version: 15.6.6, 15.7.5, 15.7.3, 15.7.5, 15.7.5, 15.7.6, 5.8 Preview 5, 5.8 Preview 4

We have been seeing many issues that involve missing MEF components, for example:

- #1826 GitHub Extension broken with VS 15.7.6 
- #1810 SetSite failed for the package [GitHubPackage], VS 15.8 Preview 4/5 
- #1812 Error On Project Load 

### Symptoms

These tend to manifest themselves as follows on the GitHub pane:
```
System.InvalidOperationException: Could not locate any instances of contract GitHub.Services.ITeamExplorerServiceHolder.
   at GitHub.VisualStudio.GitHubServiceProvider.GetService(Type serviceType) in C:\projects\visualstudio\src\GitHub.VisualStudio\Services\GitHubServiceProvider.cs:line 147
   at GitHub.VisualStudio.GitHubServiceProvider.GetService[T]() in C:\projects\visualstudio\src\GitHub.VisualStudio\Services\GitHubServiceProvider.cs:line 153
   at GitHub.VisualStudio.UI.GitHubPane.<InitializeAsync>d__12.MoveNext() in C:\projects\visualstudio\src\GitHub.VisualStudio\UI\GitHubPane.cs:line 98
```

Or like this when opening a solution that's hosted on GitHub:

image

With an entry in the ActivityLog:
```xml
<description>
SetSite failed for package [GitHubPackage]
  Source: 'System.ComponentModel.Composition'
  Description: No exports were found that match the constraint:
 ContractName=GitHub.Commands.IOpenFromUrlCommand
 RequiredTypeIdentity=GitHub.Commands.IOpenFromUrlCommand
</description>
```

### What's going on

*Update from Daniel Griffen @dgriffen on [Gitter](https://gitter.im/Microsoft/extendvs?at=5b7739225921214124814450).*

> @jcansdale I was able to reproduce this locally! I've figured out the issue and am working on how to fix it on our end or provide you with a workaround.
> @jcansdale there is a slight issue where the target version is getting the incorrect version of the running VS process and therefore excluding both your teamfoundation mef assets.
> The commit that added that to your extension was here: https://github.com/github/VisualStudio/commit/c9703cd2a284840474b191449c52bb67b181a5e4#diff-f5cce9abffe08468ca7c1c1cd08cc17f
> @jcansdale the bug is that sometimes servicehub's version is used which is not tied directly to the VS version (currently, its ~1.3 I think).

This doesn't appear to be caused by a specific version of GitHub for Visual Studio or MEF component, but seems to be a ~~random corruption of the extension cache~~ see above.

Simply clearing the MEF component cache or running `devenv.exe /updateConfiguration` won't repair the extension cache (which is usually what resolves missing MEF components).

#### Clearing the extension *and* MEF caches

The quickest way is to delete the following file and directory (where `15.0_xxxxxxxx` corresponds to a particular instance of Visual Studio):
```
%USERPROFILE%\AppData\Local\Microsoft\VisualStudio\15.0_xxxxxxxx\Extensions\extensions.en-US.cache
%USERPROFILE%\AppData\Local\Microsoft\VisualStudio\15.0_xxxxxxxx\ComponentModelCache
```

Deleting `extensions.en-US.cache` causes the extension cache to be recreated and deleting the `ComponentModelCache` directory clears the MEF component cache.

#### Reinstalling the extension

The alternative way is to uninstall and then re-install GitHub for Visual Studio.

This can be error prone when done manually because the `VSIX Installer` often appears behind other windows. It's easy to think the uninstall has completed when it hasn't even started. Perhaps we could provide a script for doing this?

### Automatic detection

Missing MEF components would be easy to detect during our package initialization. The `package did not load correctly` dialog only appears because we don't handle an exception in our package initialization. At the very least we could point users to instructions on how to repair their extension cache.

### Automatic repair

It's easier to find the relevant cache files from inside an instance of Visual Studio that it is to find the correct instance-Guid and paths from an external script. It would be straightforward to detect suspected cache corruption and offer to perform an automatic repair.

### PR DESCRIPTION
**Pull request by jcansdale**
### What's the issue

Specifying a `TargetVersion` was causing the `GitHub.TeamFoundation.15` MEF assembly to sometimes incorrectly be excluded when installed on Visual Studio 2017 (see #1859).

@dgriffen [said](https://gitter.im/Microsoft/extendvs?at=5b7af3411d3a5711b673493f):
> the bug is that sometimes servicehub's version is used which is not tied directly to the VS version (currently, its ~1.3 I think).

### What this PR does

Assuming this issue is fixed in 2019, we can assume that when the version is reported as `1.x` (instead of `15.x`) we're installing for Visual Studio 2017. Specifying a version range isn't supported when installing for Visual Studio 2015, so this isn't an issue there.

- Install for Visual Studio 2017 when version is incorrectly reported as 1.x

### How to repro

1. Install GitHub for Visual Studio v2.5.4.3349 in Visual Studio 2017
2. Launch Visual Studio 2017 (with the `GitHub` and `Team Explorer` panes closed)
3. Open the folder at: `%USERPROFILE%\AppData\Local\Microsoft\VisualStudio\15.0_xxxxxxxx`
(where `15.0_xxxxxxxx` is a unique name corresponding to the instance of Visual Studio you're running)
4. Delete the folder `ComponentModelCache`
5. Delete the file `Extensions\extensions.en-US.cache`
6. Kill the process `ServiceHub.VSDetouredHost.exe`
image
7. Open any solution that has a Git repository (you *might* see the dialog below at this point)
8. Close and re-open Visual Studio 2017
9. Open any solution that has a Git repository
10. Expect to see the following dialog:
image
11. If you've ever accidentally selected no on the above dialog, you won't see anything. 😢 Try running the `View > Other Windows > GitHub`. Expect nothing to happen.

The extensions cache (`extensions.en-US.cache`) is now corrupted. 🎉 😭  The only way to fix it is to delete `Extensions\extensions.en-US.cache` and `ComponentModelCache` and relaunch Visual Studio.

### How to test

Following on from `How to repro`...

1. Install the version from this PR
2. The extension should now be working
3. Repeat from step 2. of `How to repro`
4. Expect the extension to continue working this time

Phew! 🤞 

### Related

Fixes #1859 Corrupt extension cache and missing MEF exports
Fixes #1826 GitHub Extension broken with VS 15.7.6 
Fixes  #1810 SetSite failed for the package [GitHubPackage], VS 15.8 Preview 4/5 
Fixes  #1812 Error On Project Load

### CODE REVIEW

File: src/GitHub.VisualStudio/source.extension.vsixmanifest [hunk 1]
```diff
@@ -31,6 +31,8 @@
     <Asset Type="Microsoft.VisualStudio.MefComponent" d:Source="Project" d:ProjectName="GitHub.Services.Vssdk" Path="|GitHub.Services.Vssdk|" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" d:Source="Project" d:ProjectName="GitHub.TeamFoundation.14" TargetVersion="[14.0,15.0)" Path="|GitHub.TeamFoundation.14|" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" d:Source="Project" d:ProjectName="GitHub.TeamFoundation.15" TargetVersion="[15.0,16.0)" Path="|GitHub.TeamFoundation.15|" />
+    <!-- Sometimes servicehub's version is used (currently ~1.3) when installing for Visual Studio 2017. If the version is 1.x, assume we're installing for Visual Studio 2017.  -->
```
Comment (grokys) [inline src/GitHub.VisualStudio/source.extension.vsixmanifest]: Are we sure it's always servicehub's version that's being used, or is it just a random assembly that's loaded into the process? If the latter, then this won't fix the problem I don't think...

Comment (dgriffen) [reply src/GitHub.VisualStudio/source.extension.vsixmanifest]: The version is actually the version of the `VSDetouredHost` which can be found in `Common7\ServiceHub\Hosts\ServiceHub.Host.CLR.x86\` The reason this happens is that the VSDetouredHost loads a copy of the extension manager which may try to write to the extensions cache.

This fix *should* work but you'll want to test it to make sure you don't see the issue popping up still. I don't know enough about why you started using the target versions to say if this approach is better than just falling back to always attempting to load it.

Comment (jcansdale) [reply src/GitHub.VisualStudio/source.extension.vsixmanifest]: @dgriffen,

> I don't know enough about why you started using the target versions to say if this approach is better than just falling back to always attempting to load it.

The problem was that the MEF log ended up looking this [this](https://github.com/github/VisualStudio/files/927920/Microsoft.VisualStudio.Default.err.txt) when the extension installed normally. This made diagnosing MEF issues for our own and other extensions problematic (you needed to know which exceptions to ignore in order to find any genuine issues).

> The reason this happens is that the VSDetouredHost loads a copy of the extension manager which may try to write to the extensions cache.

When should I expect `ServiceHub.VSDetouredHost.exe` to run?

Comment (dgriffen) [reply src/GitHub.VisualStudio/source.extension.vsixmanifest]: > When should I expect ServiceHub.VSDetouredHost.exe to run?

The timing is tricky and I was never able to reliably produce a cache that wouldn't get detected as invalid by the next launch of VS. `VSDetouredHost` runs whenever someone requests a servicehub service that is not installed in the `Common7\ServiceHub\Services`. The most common time for this is project load. After you get it to regenerate the cache you will need to shutdown VS and delete the `ComponentModelCache` folder to trigger it to regenerate from the broken state.

Comment (jcansdale) [reply src/GitHub.VisualStudio/source.extension.vsixmanifest]: @dgriffen, thanks. I've added `How to repro` and `How to test` sections to the PR. I'm able to trigger it relatively consistently, even it this isn't exactly what is happening in the wild.

**Comment by jcansdale** [comment]
From @meaghanlewis 

> i was finally able to repro consistently and was just finishing testing out the build from the PR
> yeah it LGTM :thumbsup:

Merging...

**Pull request merged**